src/ray/raylet/scheduling_queue.h

```diff
@@ -40,6 +41,86 @@ enum class TaskState {
+  /// \param task The task to append to the queue.
+  /// \return Whether the append operation succeeds.
+  bool AppendTask(const TaskID &task_id, const Task &task);
+
+  /// \brief Remove a task from queue.
+  ///
+  /// \param task_id The task ID for the task to remove from the queue.
+  /// \return Whether the removal succeeds.
+  bool RemoveTask(const TaskID &task_id);
+
+  /// \brief Remove a task from queue.
+  ///
+  /// \param task_id The task ID for the task to remove from the queue.
+  /// \param removed_tasks If the task specified by task_id is successfully
+  ///  removed from the queue, the task data is appended to the vector. Can
+  ///  be a nullptr, in which case nothing is appended.
+  /// \return Whether the removal succeeds.
+  bool RemoveTask(const TaskID &task_id, std::vector<Task> *removed_tasks = nullptr);
+
+  /// \brief Check if the queue contains a specific task id.
+  ///
+  /// \param task_id The task ID for the task.
+  /// \return Whether the task_id exists in this queue.
+  bool HasTask(const TaskID &task_id) const;
+
+  /// \brief Return the task list of the queue.
+  /// \return A list of tasks contained in this queue.
+  const std::list<Task> &GetTasks() const;
+
+  /// \brief Get the total resources required by the tasks in the queue.
+  /// \return Total resources required by the tasks in the queue.
+  const ResourceSet &GetCurrentResourceLoad() const;
+
+ protected:
+  /// A list of tasks.
+  std::list<Task> task_list_;
+  /// A hash to speed up looking up a task.
+  std::unordered_map<TaskID, std::list<Task>::iterator> task_map_;
+  /// Aggregate resources of all the tasks in this queue.
+  ResourceSet current_resource_load_;
+};
+
+class ReadyQueue : public TaskQueue {
+ public:
+  /// \brief Append a task to queue.
+  ///
+  /// \param task_id The task ID for the task to append.
+  /// \param task The task to append to the queue.
+  /// \return Whether the append operation succeeds.
+  bool AppendTask(const TaskID &task_id, const Task &task);
+
+  /// \brief Remove a task from queue.
+  ///
+  /// \param task_id The task ID for the task to remove from the queue.
+  /// \return Whether the removal succeeds.
+  bool RemoveTask(const TaskID &task_id, std::vector<Task> *removed_tasks);
+
+  /// \brief Get task associated to task_id in this queue.
+  ///
+  /// \param task_id The task ID for the task to get.
+  /// \return The task corresponding to task_id.
+  const Task &GetTask(const TaskID &task_id) { return *task_map_[task_id]; }
+
+  /// \brief Get a mapping from resource shape to tasks.
+  ///
+  /// \return Mapping from resource set to task IDs with these resource requirements.
+  std::unordered_map<ResourceSet, std::unordered_set<TaskID>> &GetTasksWithResources() {
+    return tasks_with_resources_;
+  }
+
+ private:
+  /// Index from resource shape to tasks that require these resources.
+  std::unordered_map<ResourceSet, std::unordered_set<TaskID>> tasks_with_resources_;
```

Comment:
is there a reason why `unordered_set` for tasks is preferable? Why not a `std::deque`? It's `push_front` and `pop_back` have O(1) complexity. That way you'll maintain FIFO at least within each ResourceSet shape.

Reply:
That would be fabulous but cannot be done in the way the code is structured right now. It will need a refactor and should be done in a different PR (look at e.g. the SchedulingQueue::RemoveTasks function, which can remove arbitrary tasks).

We need O(1) appends and O(1) removes from the middle by key and out of the containers in the STL, only std::unordered_set seem to satisfy that.